### ISSUE TITLE
Native extensions mysql gem not supported in Puppet 3.7

### ISSUE THREAD

**Issue by vanillaslimfast** [crayfishx/hiera-mysql#9]
Unfortunately, this module does not work in Puppet/PE 3.7.x, which moved to a new JVM-based puppet master. Trying to install the gem within the context of the new puppet server fails:

```
[root@xxx puppet]# /opt/puppet/bin/puppetserver gem install hiera-mysql
Building native extensions.  This could take a while...
ERROR:  Error installing hiera-mysql:
        ERROR: Failed to build gem native extension.

    java -jar /opt/puppet/share/puppetserver/puppet-server-release.jar extconf.rb
/bin/sh: java: command not found


Gem files will remain installed in /var/opt/lib/pe-puppet-server/jruby-gems/gems/mysql-2.9.1 for inspection.
Results logged to /var/opt/lib/pe-puppet-server/jruby-gems/gems/mysql-2.9.1/ext/mysql_api/gem_make.out
```

See also this bug: https://tickets.puppetlabs.com/browse/SERVER-295

**Comment by vanillaslimfast** [comment]
Hey @crayfishx, had hoped to get this figured out on my own and submit a pull request back to you but haven't had much luck. Changing the code to use a JDBC connection is fairly straightforward, but where I'm struggling is getting it to load the correct driver or properly interpret the JDBC connection string/URL. This is mostly due to my unfamiliarity with JRuby

Let me know if you've looked into this at all or have any thoughts. I basically can get it to the point where it either errors at:

`Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Error from DataBinding 'hiera' while looking up 'pe_console_prune::ensure_prune_cron': cannot load Java class com.mysql.jdbc.Driver on node x`

Or

`Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Error from DataBinding 'hiera' while looking up 'pe_console_prune::ensure_prune_cron': No suitable driver found for jdbc:mysql://servername/puppet_config_store on node x`

Cheers

**Comment by KAZAMA7** [comment]
Hi vanillaslimfast, I am getting the similar error...that you have pasted.. could you tell me what have you done to resolve this..

**Comment by vanillaslimfast** [comment]
@KAZAMA7, I gave up. We are going to move away from using mysql in favor of straight yaml in order to support our efforts to upgrade to PE 3.7+

**Comment by crayfishx** [comment]
@dankinon @KAZAMA7 @vanillaslimfast 

I'm trying to get this working in the feature/jdbc branch in order to create a 1.0 release - old stuff is deliberatly commented out as I would like to make the backend work for both jruby and non-jruby connectors (puppet apply and hiera from the command line will require the standard connector, for example) - for now, I'd just like to get the jdbc part implemented.  After installing the jdbc-mysql gem into puppetserver it gets as far as the DriverManager.get_connection call where it fails with:

```
SERVER: Evaluation Error: Error while evaluating a Function Call, No suitable driver found for jdbc:mysql://localhost/test
```

A quick google suggests that the mysql connector jar needs to be in the load path, but the gem itself contains the jar....

```
/opt/puppetlabs/puppet/lib/ruby/gems/2.1.0/gems/jdbc-mysql-5.1.35/lib/mysql-connector-java-5.1.35-bin.jar
/opt/puppetlabs/server/data/puppetserver/jruby-gems/gems/jdbc-mysql-5.1.35/lib/mysql-connector-java-5.1.35-bin.jar
```

If anyone feels like collaborating and can get any further than this, please feel free to contribute to the feature/jdbc branch

**Comment by crayfishx** [comment]
An update on this;  I've managed to get the JDBC driver working within jruby, currently tidying things up before pushing and should have something to test later today or tomorrow.

**Comment by crayfishx** [comment]
PR #10 has workable code for JDBC

**Comment by crayfishx** [comment]
@KAZAMA7 @vanillaslimfast any suggestions / feedback before I merge this?

**Comment by vanillaslimfast** [comment]
Looks pretty slick in that it works with JDBC/JRuby and falls through to the old native implementation if on the older version.

Once this is merged I should be able to test it out in the coming weeks in our PE 3.7 test bed that is currently dead in the water

**Comment by plesher** [comment]
Crayfishx, THANK YOU very much for this update!!
PE 3.8
I was able to download the pull request, get it to build into a gem, and get it to install in the puppetserver.

The only issue I ran into was that the mysql gem ( required ) would not build into the puppetserver ( 3.8 ).
After I commented that requirement out and rebuilt the gem/package everything went in.  I manually installed the mysql gem with /opt/puppet/bin/gem to use the command line version of hiera.
All of my hiera mysql data is working correctly now.

**Comment by crayfishx** [comment]
@plesher Yes - I've been wondering about how to go about the requirement for the Gem - We could either publish a "-without-native' gem for use with puppetserver gem, but that's likely to confuse people - an easier option I think is to just publish the gem without a dependancy on jdbc or mysql and document these steps

**Comment by crayfishx** [comment]
@vanillaslimfast @plesher thanks for the feedback.  I've released this as 2.0.0 to rubygems this morning (see the readme for why it isn't 1.0.0) - please let me know if you have any issues.
